packages/viewer/src/components/viewer-transform-widget/viewer-transform-widget.tsx

```diff
@@ -0,0 +1,367 @@
+
+    this.canvasResizeObserver.disconnect();
+
+    this.hoveredChangeDisposable?.dispose();
+    this.widget?.dispose();
+  }
+
+  /**
+   * @ignore
+   */
+  @Watch('viewer')
+  protected handleViewerChanged(
+    newViewer?: HTMLVertexViewerElement,
+    oldViewer?: HTMLVertexViewerElement
+  ): void {
+    oldViewer?.removeEventListener('frameDrawn', this.handleViewerFrameDrawn);
+    oldViewer?.removeEventListener(
+      'dimensionschange',
+      this.handleViewerDimensionsChanged
+    );
+    newViewer?.addEventListener('frameDrawn', this.handleViewerFrameDrawn);
+    newViewer?.addEventListener(
+      'dimensionschange',
+      this.handleViewerDimensionsChanged
+    );
+
+    if (newViewer?.stream != null) {
+      this.controller?.dispose();
+      this.controller = new TransformController(newViewer.stream);
+    }
+  }
+
+  /**
+   * @ignore
+   */
+  @Watch('position')
+  protected handlePositionChanged(): void {
+    this.currentPosition = this.position;
+
+    this.getTransformWidget()?.updatePosition(this.currentPosition);
+
+    if (this.position == null) {
+      this.controller?.clearTransform();
+    }
+  }
+
+  public render(): h.JSX.IntrinsicElements {
+    return (
+      <Host>
+        <canvas
+          ref={(el) => {
+            this.canvasRef = el;
+          }}
+          class={classNames('widget', {
+            hovered: this.hovered != null,
+          })}
+          width={this.viewer?.viewport.width}
+          height={this.viewer?.viewport.height}
+          onPointerDown={this.handlePointerDown}
+        />
+      </Host>
+    );
+  }
+
+  private handleHoveredMeshChanged = (mesh: Mesh | undefined): void => {
+    this.hovered = mesh;
+  };
+
+  private handleViewerFrameDrawn = (): void => {
+    this.updatePropsFromViewer();
+  };
+
+  private handleViewerDimensionsChanged = (): void => {
+    const canvasRef = this.canvasRef;
+
+    if (canvasRef != null && this.viewer != null && this.viewer.frame != null) {
+      const { viewport } = this.viewer;
+
+      writeDOM(() => {
+        canvasRef.width = viewport.width;
+        canvasRef.height = viewport.height;
+      });
+
+      readDOM(() => {
+        this.canvasBounds = canvasRef.getBoundingClientRect();
+        this.getTransformWidget()?.updateDimensions(canvasRef);
+      });
+    }
+  };
+
+  private handleCanvasResize = (): void => {
+    readDOM(() => {
+      if (this.canvasRef != null) {
+        this.canvasBounds = this.canvasRef?.getBoundingClientRect();
+
+        this.getTransformWidget()?.updateDimensions(this.canvasRef);
+      }
+    });
+  };
+
+  private handleWindowPointerMove = (event: PointerEvent): void => {
+    const canvasPoint = convertPointToCanvas(
+      Point.create(event.clientX, event.clientY),
+      this.getCanvasBounds()
+    );
+    const widget = this.getTransformWidget();
+
+    if (canvasPoint != null && widget?.boundsContainsPoint(canvasPoint)) {
+      widget?.updateCursor(canvasPoint);
```

Comment:
Is there a circumstance here that we would expect the widget to be undefined? If we expect the widget to be present, I would advocate for throwing an exception any place we are trying to use the widget and it's undefined. This will make troubleshooting & debugging a lot easier. Even if it's in an `if else` statement and we include a debug log in the `else` or something. Might be useful, given the sorts of stencil related bugs we've had recently.

Reply:
That's fair, I'll take a look at adding some errors + logging to indicate what's wrong 👍